### DISCUSSION TITLE
Environment names other than [testenv:]

### DISCUSSION THREAD

**Discussion by gordonsross** [tox-dev/tox#2073] [Q&A]
I've had a good root around and not found anything relevant, but has there ever been any discussion about allowing alternatively prefixed environment names e.g. [fooenv:bar] within a tox.ini file?

**Comment by jugmac00** [comment]
What are your motivations?

**Reply by gordonsross** [reply]
My team and I really love tox. We've looked at other virtual environment management solutions (Poetry, PyFlow, Conda etc) but we keep coming back to tox. Completely understand that it's primary focus is on creating environments within which to test, but we are looking at using tox to run production code within virtual environments (Container options like Docker are not an option for us atm), and deploying ini files with [testenv:] as environment names for running this doesn't sit right. Out of scope for tox? Maybe, and I can understand that, but I'd love to hear your views.

**Reply by gaborbernat** [reply]
I don't think you can do that in tox 3 and would be mighty hard to not break things even if you try to patch it. That being said in tox 4 (see the rewrite branch) we handle such configuration file-specific options better, so in theory, you could just write a tox plugin to patch this variable https://github.com/tox-dev/tox/blob/rewrite/src/tox/config/loader/ini/replace.py#L19 (not that I'd recommend it necessarily). At the end of day testenv or prodenv, it's just semantics 👍

**Reply by imrahulmpawar** [reply]
I too think keeping that main section name in the tox.ini as a hard-coded string to "**[testenv]**" is kinda limiting the potential usefulness of this awesome package.

Tox being totally a single ini file driven package is probably the fastest virtual environment handler package out there atleast in terms of setup time.

So adding this main section name customisation option will surely open-up tox for other use-case.

One way to do this that I can think of is simply add another optional parameter(_So if that is not specified it will default to testenv_) in the root tox package/script to specify root section name that way people can provide any custom name they want like.

So new tox.ini file will look like :

[ProjectA]


[ProjectA:RunCommand1]

Script's commandline will look like : 
python -m tox --main-section ProjectA - e RunCommand1

Instead of the current syntax of 

[testenv]


[testenv:RunCommand1]
Script's commandline will look like : 
python -m tox - e RunCommand1


I hope you think about treating this as a feature.

**Comment by sanjayvjadhav** [comment]
I have used Tox in our project and its very simple and clean solution to create multiple virtual environments.

The only thing we can have in this package in a feature is to remove dependency of  hard coded section [testenv].   releasing section with [testenv] in production doesn't seems to be right.

We can give user to flexibility to create his own parent section and create child section under his own parent section instead of using current [testenv] section as a parent section. This will make tox as more generic.

Please do consider this suggestion in your next tox version  :)

**Reply by gaborbernat** [reply]
Note in tox 4 you can alter the root environment 😊 so you might replace testenv with whatever you'd like... https://github.com/tox-dev/tox/blob/rewrite/src/tox/config/source/ini.py#L72-L77 so you might get 90% there with tox 4.
